### PR TITLE
Fix inference and cleanup model loading

### LINKED ISSUE #1273: Inference is broken with forecast strategy `random`
### What happened?

Inference is broken with forecast strategy `random`

### What are the steps to reproduce the bug?

Run inference with forecast strategy `random`

### Hedgedoc link to logs and more information. This ticket is public, do not attach files directly.

_No response_

### LINKED ISSUE #1274: Duplicate load function for model
### Describe the task. It can be a feature, documentation, etc.

We currently have two `load` function for the model, one in model and one in trainer. This needs to be consolidated.

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

_No response_

### Area

- [ ] datasets, data readers, data preparation and transfer
- [ ] model
- [ ] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation

### PR DESCRIPTION
**Pull request by clessig**
## Description

Fix inference and cleanup model loading

## Issue Number

Closes #1273 
Closes #1274

## Things we tested

 - [x] Multi-GPU train default config (Sophie) [SUCCESS]
 - [x] Multi-GPU train-continue default config (Sophie) [SUCCESS]
 - [x] Single-GPU inference of Multi-GPU run default config (Sophie) [SUCCESS]
 - [ ] Multi-GPU train-continue with forecasting  (Sophie) [FAILED]

## Checklist before asking for review

-   [X] I have performed a self-review of my code
-   [ ] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [ ] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/train/trainer.py [hunk 1]
```diff
@@ -101,73 +102,17 @@ def init(self, cf: Config, devices):
         self.init_perf_monitoring()
         self.train_logger = TrainLogger(cf, config.get_path_run(self.cf))
 
-    def inference(self, cf, devices, run_id_trained, epoch):
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: Inference has been swapped with init_model_and_shard. This makes the diff a big larger than it needed to be.

File: src/weathergen/train/trainer.py [hunk 2]
```diff
@@ -260,8 +204,71 @@ def init_model_and_shard(self, cf, devices):
...
+            "batch_sampler": None,
+            "shuffle": False,
+            "num_workers": loader_num_workers,
+            "pin_memory": True,
+        }
+        self.data_loader_validation = torch.utils.data.DataLoader(
+            self.dataset, **loader_params, sampler=None
+        )
+
+        self.model, self.model_params = self.init_model_and_shard(
```
Comment (clessig) [inline src/weathergen/train/trainer.py:259]: model.load() has been replaced with the function used for train/train_continue.

File: src/weathergen/train/trainer.py [hunk 3]
```diff
@@ -302,28 +310,17 @@ def run(self, cf, devices, run_id_contd=None, epoch_contd=None):
             self.dataset_val, **loader_params, sampler=None
         )
 
-        self.model, self.model_params = self.init_model_and_shard(cf, devices)
-
-        if run_id_contd is None:
```
Comment (clessig) [inline src/weathergen/train/trainer.py:307]: has been moved to init_model_and_shard to be able to use it in run() and in inference()

File: src/weathergen/train/trainer.py [hunk 4]
```diff
@@ -794,89 +790,49 @@ def load_model(self, run_id: str, epoch=-1):
...
+        # warn about difference in checkpoint and model
+        if len(mkeys) == 0 and len(ukeys) == 0:
+            logger.info(f"Checkpoint {filename} loaded successfully with all weights matching.")
         if len(mkeys) > 0:
             logger.warning(f"Missing keys when loading model: {mkeys}")
-
         if len(ukeys) > 0:
             logger.warning(f"Unused keys when loading model: {mkeys}")
 
-    def load_optim(self):
```
Comment (clessig) [inline src/weathergen/train/trainer.py:843]: Removed since it's not used.

**Comment by sophie-xhonneux** [comment]
Tested several things, everything worked except train_continue moving from masking to forecasting

**Review by sophie-xhonneux — APPROVED** [review]
